src/mlpack/methods/ann/layer/lambda_map_reduce_impl.hpp

```diff
@@ -0,0 +1,416 @@
+  other.layerGradients.clear();
+}
+
+template<typename ReductionRuleType, typename MatType>
+LambdaMapReduceType<ReductionRuleType, MatType>& LambdaMapReduceType<ReductionRuleType, MatType>::operator=(const LambdaMapReduceType& other)
+{
+  if (this != &other)
+  {
+    Layer<MatType>::operator=(other);
+
+    network.clear();
+    layerOutputs.clear();
+    layerDeltas.clear();
+    layerGradients.clear();
+
+    inSize = other.inSize;
+    totalInputSize = other.totalInputSize;
+    totalOutputSize = other.totalOutputSize;
+    reductionRule = other.reductionRule;
+
+    layerOutputMatrix = other.layerOutputMatrix;
+    layerDeltaMatrix = other.layerDeltaMatrix;
+
+    for (size_t i = 0; i < other.network.size(); ++i)
+      network.push_back(other.network[i]->Clone());
+
+    // Ensure that the aliases for layers during passes have the right size.
+    layerOutputs.resize(network.size(), MatType());
+    layerDeltas.resize(network.size(), MatType());
+    layerGradients.resize(network.size(), MatType());
+  }
+
+  return *this;
+}
+
+template<typename ReductionRuleType, typename MatType>
+LambdaMapReduceType<ReductionRuleType, MatType>& LambdaMapReduceType<ReductionRuleType, MatType>::operator=(LambdaMapReduceType&& other)
+{
+  if (this != &other)
+  {
+    Layer<MatType>::operator=(other);
+
+    layerOutputs.clear();
+    layerDeltas.clear();
+    layerGradients.clear();
+
+    inSize = std::move(other.inSize);
+    totalInputSize = std::move(other.totalInputSize);
+    totalOutputSize = std::move(other.totalOutputSize);
+    reductionRule = std::move(other.reductionRule);
+
+    network = std::move(other.network);
+
+    layerOutputs.resize(network.size(), MatType());
+    layerDeltas.resize(network.size(), MatType());
+    layerGradients.resize(network.size(), MatType());
+
+    other.layerOutputs.clear();
+    other.layerDeltas.clear();
+    other.layerGradients.clear();
+  }
+
+  return *this;
+}
+
+template<typename ReductionRuleType, typename MatType>
+void LambdaMapReduceType<ReductionRuleType, MatType>::Forward(
+    const MatType& input, MatType& output)
+{
+  Forward(input, output, 0, network.size() - 1);
+}
+
+template<typename ReductionRuleType, typename MatType>
+void LambdaMapReduceType<ReductionRuleType, MatType>::Forward(
+    const MatType& input,
+    MatType& output,
+    const size_t start,
+    const size_t end)
+{
+  // Make sure training/testing mode is set right in each layer.
+  for (size_t i = 0; i < network.size(); ++i)
+    network[i]->Training() = this->training;
+
+  // Note that we use `output` for the last layer; layerOutputs is only used for
+  // intermediate values between layers.
+  if ((end - start) > 0)
+  {
+    // Initialize memory for the forward pass (if needed).
+    InitializeForwardPassMemory(input.n_cols);
+
+    for (size_t i = start; i <= end; i++)
+      network[i]->Forward(input, layerOutputs[i]);
+
+		// Reduce the outputs to single output.
+    reductionRule.Reduce(layerOutputs, output);
```

Comment:
Technically another more memory-efficient way to do this would be to only have one temporary `layerOutput` (since they should all be the same size), and overwrite it in each call to `network[i]->Forward()`, and then add that to `output`.  I think that should work for all reduction types.

Reply:
That will be true for add or product. But I need to make this generalized to adapt Concat fn too.

Reply:
Ah, good point, each individual layer to be concatenated could have different dimensions along the axis of concatenation.